Scripts/Runtime/Data/Collections/UnsafeTypedStream.cs

```diff
@@ -234,7 +236,7 @@ public void Dispose()
         /// Clears all data in the collection
         /// </summary>
         [WriteAccessRequired]
-        public void Clear()
+        public void ClearAndReduceCapacity()
```

Comment:
`List<T>` has `TrimExcess()`

Should we follow that lead and have excess trimming its own method?
So you'd call `Clear()` then `TrimExcess()` if you wanted to collapse the capacity down.

Or are there optimizations in assuming that you're doing a clear at the same time as the trim?

Reply:
It took me a while to figure out how to do this without adding a bunch of extra data to monitor state but I got it. 

This is a good idea. Pasting here because you won't actually see it until the next PR. 

```
        /// <summary>
        /// Frees memory for blocks of elements that are not being used.
        /// </summary>
        /// <remarks>
        /// Because this is a block based collection, trimming will not be a tight fit.
        /// Ex. 10 blocks allocated, 4 elements per block, 40 elements capacity total.
        /// A <see cref="Clear()"/> is called and then 5 elements are written and then <see cref="TrimExcess"/> is
        /// called.
        /// The first block of 4 elements will stay. The 5th element will keep the second block around. The remaining
        /// 8 blocks will be freed.
        /// In the end, 2 blocks will remain, 5 elements in the collection, 8 elements capacity.
        /// </remarks>
        [WriteAccessRequired]
        public void TrimExcess()
        {
            if (!IsCreated)
            {
                return;
            }

            Allocator blockAllocator = m_BufferInfo->BlockAllocator;

            for (int i = 0; i < m_BufferInfo->LaneCount; ++i)
            {
                LaneInfo* lane = m_BufferInfo->LaneInfos + i;
                
                //If our current writing block is null, we can just skip, there's nothing to free
                if (lane->CurrentWriterBlock == null)
                {
                    continue;
                }
                
                //If we have no elements at all in the lane (Clear was called), then we want to start with clearing
                //the current writer block. If there is anything written, (We're calling this to free up excess memory),
                //then the current writer block is still valid and we only want to free blocks after.
                BlockInfo* block = (lane->Count == 0) ? lane->CurrentWriterBlock : lane->CurrentWriterBlock->Next;
                while (block != null)
                {
                    BlockInfo* currentBlock = block;
                    block = currentBlock->Next;

                    UnsafeUtility.Free(currentBlock->Data, blockAllocator);
                    UnsafeUtility.Free(currentBlock, blockAllocator);
                }
            }
        }
```